src/JustEat.StatsD/TagsFormatters/StatsDTagsFormatter.cs

```diff
@@ -0,0 +1,172 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Text;
+using JustEat.StatsD.Buffered;
+
+namespace JustEat.StatsD.TagsFormatters
+{
+    /// <summary>
+    /// A template class to format StatsD tags with configured values.
+    /// </summary>
+    public abstract class StatsDTagsFormatter : IStatsDTagsFormatter
+    {
+        private const int SafeUdpPacketSize = 512;
+
+        [ThreadStatic]
+        private static char[]? _buffer;
+        private static char[] Buffer => _buffer ??= new char[SafeUdpPacketSize];
+
+        private readonly char[] _prefix;
+        private readonly char[] _suffix;
+        private readonly char[] _tagsSeparator;
+        private readonly char[] _keyValueSeparator;
+        private readonly int _prefixSize;
+        private readonly int _suffixSize;
+        private readonly int _tagsSeparatorSize;
+        private readonly int _keyValueSeparatorSize;
+        
+        /// <summary>
+        /// Initializes a new instance of the <see cref="StatsDTagsFormatter"/> class.
+        /// </summary>
+        /// <param name="configuration">The configuration.</param>
```

Comment:
Add an exception tag that documents the `ArgumentNullException` if `configuration` is `null`.

Reply:
Done!